File: python/src/code_smells/oop_abusers/refused_bequest.py
"""
Refused Bequest Code Smell Example

A subclass inherits methods from a parent class but doesn't use them
or overrides them with empty implementations.

This indicates incorrect use of inheritance - the subclass doesn't need
all the parent's functionality.

Refactoring suggestion: Replace Inheritance with Delegation, Push Down Method
"""


class BaseController:
    """Base controller with full functionality"""

    def start(self) -> None:
        print('starting')

    def stop(self) -> None:
        print('stopping')

    def reset(self) -> None:
        print('resetting')


class ReadOnlyController(BaseController):
    """
    Refused bequest: inherits start() and stop() but refuses to use them.

    This class overrides methods with empty implementations because it
    doesn't need that functionality. This suggests inheritance is not
    the right relationship here.
    """

    def start(self) -> None:
        pass  # Refuses to implement

    def stop(self) -> None:
        pass  # Refuses to implement


def demo_refused_bequest(readonly: bool) -> None:
    controller: BaseController = ReadOnlyController() if readonly else BaseController()
    controller.start()
    controller.stop()

File: go/code_smells/oop_abusers/refused_bequest.go
package oop_abusers

import "fmt"

// Code smell: Refused Bequest [Herencia rechazada].
// ReadOnlyController implementa la interfaz Controller pero no puede (o no debe)
// implementar comportamientos que modifiquen el estado, dejando métodos vacíos.

// Ejercicio: Añade un método de ciclo de vida Pause a la interfaz Controller
// y haz que Start y Stop sean obligatorios con lógica real.

// ReadOnlyController se verá forzado a implementar métodos que no tienen sentido
// para su propósito, lo que te obligará a lanzar excepciones o dejar implementaciones
// vacías que violan el Principio de Sustitución de Liskov.

type Controller interface {
	Start()
	Stop()
}

type Resettable interface {
	Reset()
}

type BaseController struct{}

func (b *BaseController) Start() {
	fmt.Println("starting")
}

func (b *BaseController) Stop() {
	fmt.Println("stopping")
}

func (b *BaseController) Reset() {
	fmt.Println("resetting")
}

type ReadOnlyController struct{}

func (r *ReadOnlyController) Start() {
	// Método vacío - refused bequest
}

func (r *ReadOnlyController) Stop() {
	// Método vacío - refused bequest
}

func DemoRefusedBequest(readonly bool) {
	var controller Controller
	if readonly {
		controller = &ReadOnlyController{}
	} else {
		controller = &BaseController{}
	}
	controller.Start()
	controller.Stop()
}
